### ISSUE TITLE
[BUG] pypwalker failed to load the visual in Streamlit 

### ISSUE THREAD

**Issue by dickhfchan** [Kanaries/pygwalker#604]
**Describe the bug**
After selecting a field in either x or y axis, pypwalker showed the visual but very quickly the selection was cleared with blank visual screen.

**To Reproduce**
1. Copy the demo example of pypwalker for streamlit
2. Run the codes
3. Select any field to the axis
4. See error as described above

**Versions**
- pygwalker version: pygwalker==0.4.9.4
- python version 3.12.4
- browser latest Chrome
- streamlit==1.37.1

<img width="1280" alt="image" src="https://github.com/user-attachments/assets/b2816582-ecb7-4b2d-a9cd-73e3fa88e07d">

**Comment by longxiaofei** [comment]
@dickhfchan  Thanks for your feedback.

Since there are many sample codes for pygwalker in Streamlit, I need to confirm which demo code you are running.

**Comment by ubercylon8** [comment]
Showing exactly the same issue on my side. This is my very simple code:
import streamlit as st
import pandas as pd
from pygwalker.api.streamlit import StreamlitRenderer

# Title of the app
st.set_page_config(layout="wide")
st.title("JSON to Pandas DataFrame and Visualization")

# File uploader
uploaded_file = st.file_uploader("Choose a JSON file", type="json")

if uploaded_file is not None:
    # Read the JSON file into a Pandas DataFrame
    df = pd.read_json(uploaded_file)
    
    # Convert datetime columns to strings
    for col in df.select_dtypes(include=['datetime64[ns]', 'datetime64[ns, UTC]']).columns:
        df[col] = df[col].astype(str)
    
    # Pygwalker 
    pyg_app = StreamlitRenderer(df)
 
    pyg_app.explorer()

**Comment by longxiaofei** [comment]
oh! I reproduced it on cloud streamlit, I will find out the cause and fix it.

**Comment by longxiaofei** [comment]
Hi @ubercylon8 , you can use the following code to complete your requirements temporary.

Since Pygwalker 0.4.9.4 uses native Streamlit components for rendering, when the spec changes, it passes the spec value to the Python side, which triggers a rerun of streamlit and causes this bug.

I will try to add this caching in a more implicit way.


```python
import streamlit as st
import pandas as pd
from pygwalker.api.streamlit import StreamlitRenderer

st.set_page_config(layout="wide")
st.title("JSON to Pandas DataFrame and Visualization")


@st.cache_resource
def get_pyg_renderer(df: "pd.DataFrame") -> "StreamlitRenderer":
    return StreamlitRenderer(df)


uploaded_file = st.file_uploader("Choose a JSON file", type="json")

if uploaded_file is not None:
    cur_df = pd.read_json(uploaded_file)

    for col in cur_df.select_dtypes(include=['datetime64[ns]', 'datetime64[ns, UTC]']).columns:
        cur_df[col] = cur_df[col].astype(str)

    pyg_app = get_pyg_renderer(cur_df)

    pyg_app.explorer()
```

**Comment by longxiaofei** [comment]
fixed it(`pygwalker==0.4.9.5`)
